src/test/java/com/google/sps/servlets/NewRequestServletTest.java

```diff
@@ -95,23 +124,33 @@ public void testStartTimeAfterEndTime() throws IOException, ServletException {
     verify(response).sendError(HttpServletResponse.SC_BAD_REQUEST, "Please enter a valid time!");
   }
   /**
-   * Tests that when correct data is sent to the servlet, no errors are sent.
-   * This test will be added when Firebase Credentials will be set
+   * Tests that when correct data is sent to the servlet, no BAD_REQUEST errors are sent. 
    */
-//   @Test
-//   public void testValidData() throws IOException, ServletException {
-//     when(request.getParameter("delivery-date")).thenReturn("2020-09-26");
-//     when(request.getParameter("start-time")).thenReturn("05:15");
-//     when(request.getParameter("end-time")).thenReturn("15:10");
-//     when(request.getParameter("max-stops")).thenReturn("1");
-//     when(request.getParameter("timezone-offset")).thenReturn("180");
-
-//     NewRequestServlet requestServlet = new NewRequestServlet();
-//     requestServlet.doPost(request, response);
-//     // the date is valid, thus "Please enter a valid date!" shouldn't be sent
-//     verify(response, times(0)).sendError(HttpServletResponse.SC_BAD_REQUEST, "Please enter a valid date!");
-//     verify(response, times(0)).sendError(HttpServletResponse.SC_BAD_REQUEST, "Please enter a valid time!");
-//     verify(response, times(0)).sendError(HttpServletResponse.SC_BAD_REQUEST, "Please enter between 1 and 25 stops!");
-//     //TODO[ak47na]: check that the delivery request is added to Mock datastore
-//   }
+  @Test
+  public void testValidData() throws IOException, ServletException, FirebaseAuthException {
+    when(request.getParameter("delivery-date")).thenReturn("2020-09-26");
+    when(request.getParameter("start-time")).thenReturn("05:15");
+    when(request.getParameter("end-time")).thenReturn("15:10");
+    when(request.getParameter("max-stops")).thenReturn("1");
+    when(request.getParameter("timezone-offset-minutes")).thenReturn("180");
+    when(request.getParameter("start-address")).thenReturn("277 Bedford Ave, Brooklyn, NY 11211, USA");
+
+    when(response.getWriter()).thenReturn(new PrintWriter(System.out));
+
+    NewRequestServlet requestServlet = spy(new NewRequestServlet());
+
+    when(request.getParameter("idToken")).thenReturn(idToken);
+    when(firebaseAuth.getUserIdFromIdToken(idToken)).thenReturn(userId);
+    
+    requestServlet.setFirebaseAuth(firebaseAuth);
+    requestServlet.doPost(request, response);
+
+    DatastoreService ds = DatastoreServiceFactory.getDatastoreService();
+    // The user is a new user and it is added to datastore.
+    Query userQuery = new Query("UserData")
+          .setFilter(new Query.FilterPredicate("uid", Query.FilterOperator.EQUAL, userId));
+    assertEquals(1, ds.prepare(userQuery).countEntities(FetchOptions.Builder.withLimit(10)));
+    // TODO[ak47na]: update test to check that the delivery request is added after 
```

Comment:
Test should be super focused and minimal.

You should not have a single test that tests "valid data". You should have multiple very specific ones, e.g.:

- testNewRequestsAutomaticallyCreatesCourierProfile
- testNewRequestEnqueuesNewDeliveryRequestInDataStore
- testNewRequestFailsForUnauthorizedUser

etc.

Reply:
I changed the name from testValidData() to testNewRequestsAutomaticallyCreatesCourierProfile() and I will add  

- testNewRequestEnqueuesNewDeliveryRequestInDataStore

- testNewRequestFailsForUnauthorizedUser